packages/actor-query-operation-path-zero-or-more/lib/ActorQueryOperationPathZeroOrMore.ts

```diff
@@ -16,14 +19,59 @@ export class ActorQueryOperationPathZeroOrMore extends ActorAbstractPath {
   }
 
   public async runOperation(path: Algebra.Path, context: ActionContext): Promise<IActorQueryOperationOutputBindings> {
+    context = this.isPathArbitraryLengthDistinct(context);
+    if (context.get('isPathArbitraryLengthDistinct')) {
+      return ActorQueryOperation.getSafeBindings(await this.mediatorQueryOperation.mediate({
+        operation: ActorAbstractPath.FACTORY.createDistinct(path),
+        context,
+      }));
+    }
+
     const predicate = <Algebra.ZeroOrMorePath> path.predicate;
 
     const sVar = path.subject.termType === 'Variable';
     const oVar = path.object.termType === 'Variable';
 
     if (sVar && oVar) {
-      throw new Error('ZeroOrMore path expressions with 2 variables not supported yet');
-    } else if (!sVar && !oVar) {
+      // Get all the results of subjects, then fill in first variable for those
+      const predVar = this.generateVariable(path);
+      const single = ActorAbstractPath.FACTORY.createUnion(
+        ActorAbstractPath.FACTORY.createPattern(path.subject, predVar, path.object, path.graph),
+        ActorAbstractPath.FACTORY.createPattern(path.object, predVar, path.subject, path.graph),
+      );
```

Comment:
Why is this union here? Am I wrong in thinking that this entire block should be identical to the one in the OneOrMore operator?

Reply:
In the zeroOrMore, every namedNode in the dataset exists in a binding to itself, so this is how I fetch all entities in the dataset (all subjects + all objects) because they all apply to the query using the path 0 times, this was not necessary in the oneOrMore.

Reply:
But this block corresponds to the case where both subject and object are a variable, so both queries in the union would generate identical results. Those specific cases would be caught when you call this function recursively below with an instantiated variable.

Reply:
I'm not sure that will work. A namedNode that only occurs as object in the dataset, not even with the requested predicate in our zeroOrMore should also be included. A possibility is without a union, but then extracting the returned bindings in two seperate results (the subject and the object).

Reply:
Ah true, both subject and object would need to be checked for this specific instance (potentially, would have to think a bit more about the details to be sure). But this is not the way to do it from what I can see. To convert to SPARQL what is happening here, the input pattern is `?s :p+ ?o.`. In the lines above you convert this to `{ ?s :p+ ?o. } UNION { ?o :p+ ?s. }`. Since variable names don't matter, this is a union of 2 identical streams.

Reply:
Actually I convert it to `{ ?s ?p ?o} UNION { ?o ?p ?s }`. The question is actually: "How to get put a query that returns a binding of size 1 with every namedNode in the dataset?". This was where I came up with, but it's indeed not very good.

Reply:
Do you have an alternative to get al those bindings of size 1?

Reply:
Ah yes, I remember now why this part wasn't implemented, because requesting a stream of all triples to an endpoint is not really ideal. You could do it with a single `?s ?p ?o` pattern and then apply a transform iterator to it that checks both subject and object, which would "only" have to create 1 stream of all triples. Another optimization could be to first return all the results for if the query was `?s :p+ ?o.` (instead of `*`) and only append those single bindings afterwards since those are going to be way more inefficient to calculate.